### PR TITLE
feat(modal): on-before-ok return Promise<any>

### PR DESCRIPTION
**Pull request by czq297297**
## Types of changes

- [x] New feature
- [ ] Bug fix
- [ ] Enhancement
- [ ] Component style change
- [ ] Typescript definition change
- [ ] Documentation change
- [ ] Coding style change
- [ ] Refactoring
- [ ] Test cases
- [ ] Continuous integration
- [ ] Breaking change
- [ ] Others

## Changelog

| Component | Changelog(CN) | Changelog(EN) | Related issues |
| --------- | ------------- | ------------- | -------------- |
|           |               |               |                |

## Checklist:

- [ ] Test suite passes (`npm run test`)
- [ ] Provide changelog for relevant changes (e.g. bug fixes and new features) if applicable.
- [ ] Changes are submitted to the appropriate branch (e.g. features should be submitted to `feature` branch and others
  should be submitted to `main` branch)

### CODE REVIEW

**Comment by czq297297** [comment]
如果没问题  我将给drawer popconfirm的onbeforeok都加上

File: packages/web-vue/components/modal/modal.vue [hunk 1]
```diff
@@ -332,7 +332,7 @@ export default defineComponent({
      */
     onBeforeOk: {
       type: [Function, Array] as PropType<
-        (done: (closed: boolean) => void) => void | boolean
+        (done: (closed: boolean) => void) => void | boolean | Promise<any>
```
Comment (tazyong) [inline packages/web-vue/components/modal/modal.vue:335]: 返回类型应为 `Promise<void | boolean>` 更为合适，返回 `Promise<false>` 时应当阻止对话框关闭

File: packages/web-vue/components/modal/modal.vue [hunk 2]
```diff
@@ -543,6 +543,15 @@ export default defineComponent({
 
           if (isBoolean(result)) {
             resolve(result);
+          } else if (isPromise(result)) {
```
Comment (tazyong) [inline packages/web-vue/components/modal/modal.vue:546]: handleOk 改为这样：

```ts
const handleOk = async () => {
  const currentPromiseNumber = promiseNumber;
  const closed = await new Promise<boolean>(
    // eslint-disable-next-line no-async-promise-executor
    async (resolve) => {
      if (isFunction(props.onBeforeOk)) {
        let result = props.onBeforeOk(resolve);
        if (isPromise(result) || !isBoolean(result)) {
          _okLoading.value = true;
        }
        if (isPromise(result)) {
          result = await result;
        }
        if (isBoolean(result)) {
          resolve(result);
        }
      } else {
        resolve(true);
      }
    }
  );

  if (currentPromiseNumber === promiseNumber) {
    if (closed) {
      emit('ok');
      close();
    } else if (_okLoading.value) {
      _okLoading.value = false;  // _okLoading 调整至此处是因为上面的 close 函数中已经处理了这个状态，没必要重复处理
    }
  }
};
```

**Review by tazyong — COMMENTED** [review]
处理规则修改

**Pull request closed without merging**